### PR TITLE
[Advanced dreambooth lora] adjustments to align with canonical script

### PR DESCRIPTION
**Pull request by linoytsaban**
minor changes to align with canonical script - such as validation logging etc

### CODE REVIEW

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 1]
```diff
@@ -571,7 +636,7 @@ def parse_args(input_args=None):
     parser.add_argument(
         "--optimizer",
         type=str,
-        default="adamW",
+        default="AdamW",
         help=('The optimizer type to use. Choose between ["AdamW", "prodigy"]'),
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:640]: you can set this like:

```py
available_optimizers = ['AdamW', 'prodigy']
parser.add_argument(
...
    choices=available_optimizers,
    help=f'The optimizer type to use. Choose between {available_optimizers}'
)
```

Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:640]: which ensures the user doesn't select a nonexistent option

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 2]
```diff
@@ -1178,13 +1239,12 @@ def encode_prompt(text_encoders, tokenizers, prompt, text_input_ids_list=None):
...
         prompt_embeds = text_encoder(
-            text_input_ids.to(text_encoder.device),
-            output_hidden_states=True,
+            text_input_ids.to(text_encoder.device), output_hidden_states=True, return_dict=False
         )
 
         # We are only ALWAYS interested in the pooled output of the final text encoder
         pooled_prompt_embeds = prompt_embeds[0]
-        prompt_embeds = prompt_embeds.hidden_states[-2]
+        prompt_embeds = prompt_embeds[-1][-2]
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1247]: this is no longer necessary as of pytorch 2.3 but i guess the version we rely on so far isn't new enough

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 3]
```diff
@@ -1200,9 +1260,16 @@ def main(args):
...
         )
+
     if args.do_edm_style_training and args.snr_gamma is not None:
         raise ValueError("Min-SNR formulation is not supported when conducting EDM-style training.")
 
+    if torch.backends.mps.is_available() and args.mixed_precision == "bf16":
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        raise ValueError(
+            "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+        )
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1271]: equally this is actually supported in pytorch 2.3 now, with a preference toward 2.4 on mps

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 4]
```diff
@@ -1778,7 +1848,8 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
     if freeze_text_encoder and not train_dataset.custom_instance_prompts:
         del tokenizers, text_encoders
         gc.collect()
-        torch.cuda.empty_cache()
+        if torch.cuda.is_available():
+            torch.cuda.empty_cache()
```
Comment (bghira) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1852]: torch.mps has an equivalent to this, if you're interested. though it seems out of scope for these adjustments

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 5]
```diff
@@ -223,10 +224,73 @@ def save_model_card(
 
 Special VAE used for training: {vae_path}.
 
+Please adhere to the licensing terms as described [here](https://huggingface.co/playgroundai/playground-v2.5-1024px-aesthetic/blob/main/LICENSE.md).
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:227]: Yes, I think this should be conditional indeed.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 1 above)
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:641]: Yes, +1 to this suggestion. `choices` greatly simplifies things.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 6]
```diff
@@ -597,7 +663,7 @@ def parse_args(input_args=None):
     parser.add_argument("--prodigy_decouple", type=bool, default=True, help="Use AdamW style decoupled weight decay")
     parser.add_argument("--adam_weight_decay", type=float, default=1e-04, help="Weight decay to use for unet params")
     parser.add_argument(
-        "--adam_weight_decay_text_encoder", type=float, default=None, help="Weight decay to use for text_encoder"
+        "--adam_weight_decay_text_encoder", type=float, default=1e-03, help="Weight decay to use for text_encoder"
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:666]: Is this not slightly backwards harming? Earlier, if folks would have used the script, the `adam_weight_decay_text_encoder` was always None, so it didn't have any effect. Now, if they continue to resume from that trained checkpoint, things will be different, no? 

So, with that said, I think it's best to not change it. 

WDYT?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 3 above)
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1272]: Okay I think for the versions to get a bit more stable till we updated that.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 7]
```diff
@@ -1289,10 +1361,10 @@ def main(args):
         if args.output_dir is not None:
             os.makedirs(args.output_dir, exist_ok=True)
 
-        model_id = args.hub_model_id or Path(args.output_dir).name
-        repo_id = None
         if args.push_to_hub:
-            repo_id = create_repo(repo_id=model_id, exist_ok=True, token=args.hub_token).repo_id
+            repo_id = create_repo(
+                repo_id=args.hub_model_id or Path(args.output_dir).name, exist_ok=True, token=args.hub_token
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1366]: If `hub_token` is not being used let's remove it from the CLI args, altogether.

**Review by sayakpaul — COMMENTED** [review]
Left some minor comments.  But the refactor looks very nice to me!

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 1 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:649]: That would be a breaking change though, because atm - if a user specifies an optimizer other than Adam/prodigy, we log a warning default to Adam (see below)

```
# Optimizer creation
    if not (args.optimizer.lower() == "prodigy" or args.optimizer.lower() == "adamw"):
        logger.warning(
            f"Unsupported choice of optimizer: {args.optimizer}.Supported optimizers include [adamW, prodigy]."
            "Defaulting to adamW"
        )
        args.optimizer = "adamw"
```

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py (hunk 7 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1374]: Hmm I'm not sure I follow? it is still in use

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 8]
```diff
@@ -2332,105 +2365,59 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
-            train_text_encoder=args.train_text_encoder,
-            train_text_encoder_ti=args.train_text_encoder_ti,
-            token_abstraction_dict=train_dataset.token_abstraction_dict,
-            instance_prompt=args.instance_prompt,
-            validation_prompt=args.validation_prompt,
-            repo_folder=args.output_dir,
-            vae_path=args.pretrained_vae_model_name_or_path,
-        )
         if args.push_to_hub:
+            save_model_card(
```
Comment (apolinario) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2408]: Move this outsie of push_to_hub so folks learn how to load the textual inversion embeddings

**Review by sayakpaul — APPROVED** [review]

**Comment by apolinario** [comment]
Is the failing test unrelated?

**Comment by sayakpaul** [comment]
100 percent not. Let’s merge once this CI run round is done.

**Comment by linoytsaban** [comment]
let me make a tiny last change first

**Comment by sayakpaul** [comment]
"real artists ship"

**Comment by TornjV** [comment]
Seems like the changes made in this one broke the fix issued here https://github.com/huggingface/diffusers/pull/6464
